### PR TITLE
Fix ButtonClicker random crash issue [ the crash happens on 64 bit gpg sdk ]

### PR DESCRIPTION
**Pull request by ggfan**
1) initialize the jui_helper::JUIWindow::init() before trying to display Toast.  GetInstance()->GetContext() needs it inside InitToast()
2)Synchronize UI init with Google Play Games authentication UI with a mutex

### CODE REVIEW

File: samples-android/ButtonClicker/jni/Application.mk [hunk 1]
```diff
@@ -1,7 +1,7 @@
-APP_PLATFORM := android-9
```
Comment (claywilkinson) [inline samples-android/ButtonClicker/jni/Application.mk]: can we keep this at 9?

File: samples-android/ButtonClicker/jni/Application.mk [hunk 2]
```diff
@@ -1,7 +1,7 @@
...
+APP_PLATFORM := android-22
 #APP_ABI := all
-APP_ABI := armeabi-v7a
-#,arm64-v8a
+#APP_ABI := armeabi-v7a
+APP_STL := arm64-v8a
```
Comment (claywilkinson) [inline samples-android/ButtonClicker/jni/Application.mk]: there is no public 64bit SDK  yet - can we keep it 32bit?

File: samples-android/ButtonClicker/jni/ButtonClickerNativeActivity.cpp [hunk 3]
```diff
@@ -105,15 +105,14 @@ void Engine::InitGooglePlayGameServices() {
...
  */
 void Engine::OnAuthActionStarted(gpg::AuthOperation op) {
+  startup_mutex_.lock();
   if (!initialized_resources_) return;
+
   ndk_helper::JNIHelper::GetInstance()->RunOnUiThread([this, op]() {
     EnableUI(false);
-    authorizing_ = true;
     if (op == gpg::AuthOperation::SIGN_IN) {
-      LOGI("Signing in to GPG");
```
Comment (claywilkinson) [inline samples-android/ButtonClicker/jni/ButtonClickerNativeActivity.cpp]: can you keep the log messages?  They help diagnose issues which there always are with authentication.

File: samples-android/ButtonClicker/jni/ButtonClickerNativeActivity.h [hunk 4]
```diff
@@ -147,13 +147,13 @@ class Engine : public gpg::IRealTimeEventListener {
...
   int32_t score_counter_;   // Score counter of local player
   bool playing_;            // Am I playing a game?
-  bool authorizing_;        // Am I signing in to gpg service?
   std::string self_id_;     // Local player's ID
   double start_time_;       // Game start time
 
   // synchronization primitive to synchronize
   // UIThread, Timer thread and gpg callback thread
   mutable std::mutex mutex_;
+  mutable std::mutex startup_mutex_;
```
Comment (claywilkinson) [inline samples-android/ButtonClicker/jni/ButtonClickerNativeActivity.h]: can you add a comment on why 2 mutex are needed?  It seems like 1 should be ok?

File: samples-android/ButtonClicker/jni/ButtonClickerNativeActivity_Engine.cpp [hunk 5]
```diff
@@ -61,10 +60,12 @@ void Engine::UnloadResources() { renderer_.Unload(); }
  */
 int Engine::InitDisplay(const int32_t cmd) {
   if (!initialized_resources_) {
+    startup_mutex_.lock();
     gl_context_->Init(app_->window);
     InitUI();
     LoadResources();
     initialized_resources_ = true;
+    startup_mutex_.unlock();
```
Comment (claywilkinson) [inline samples-android/ButtonClicker/jni/ButtonClickerNativeActivity_Engine.cpp]: can you catch exceptions to make sure the mutex gets unlocked?

File: samples-android/ButtonClicker/jni/Application.mk (hunk 2 above)
Comment (ggfan) [inline samples-android/ButtonClicker/jni/Application.mk]: that is true: will keep it 32 bit only.

File: samples-android/ButtonClicker/jni/ButtonClickerNativeActivity.cpp (hunk 3 above)
Comment (ggfan) [inline samples-android/ButtonClicker/jni/ButtonClickerNativeActivity.cpp]: removed a couple of LOG messages from sample to make it clean.  when UI is up, it is displayed UI page at the bottom in status_text. so we could make it  a little clean in logcat

File: samples-android/ButtonClicker/jni/ButtonClickerNativeActivity.h (hunk 4 above)
Comment (ggfan) [inline samples-android/ButtonClicker/jni/ButtonClickerNativeActivity.h]: this one is just for start up code: taking conservative way not to make a big change to the existing code.

**Comment by ggfan** [comment]
there are 2 scenarios: 
  1) if the app is not logged in last time, there is no impact -- user will need to log in explicitly; the lock does not matter.
 2) if app already logged in last time, when app starts up, gpg will try silent login, and it will try to display pop-up UI. this is the period that we dedicate the UI to gpg so it could display its login UI, not complicate it with UI thread  sharing the display with it.

For the above case 1), you could not initDisplay after authentication: I think user need to click/tap a button to start login process
For the above case 2), we do not have much control of the gpg because gpg init is inside android_main() loop, it is a head of UI initialization.

It might make sense to force gpg init after UI is initialized, but that would need somewhat re-layout / evaluate overall UI and gpg login code. that would need more experiment. Current fix will make sure we are forcing the thing done in the thread that originated the action [ I think this simplify things ]

File: samples-android/ButtonClicker/jni/ButtonClickerNativeActivity_Engine.cpp (hunk 5 above)
Comment (ggfan) [inline samples-android/ButtonClicker/jni/ButtonClickerNativeActivity_Engine.cpp]: this is good point.  if there is exception in native code, it crashes and there is not any catcher here. I think java code handles lot of exceptions; in side native code, I do not see too much yet.

File: samples-android/ButtonClicker/jni/Application.mk (hunk 1 above)
Comment (ggfan) [inline samples-android/ButtonClicker/jni/Application.mk]: yes, sure

File: samples-android/ButtonClicker/jni/ButtonClickerNativeActivity.cpp (hunk 3 above)
Comment (claywilkinson) [inline samples-android/ButtonClicker/jni/ButtonClickerNativeActivity.cpp]: But when there is a problem, we ask for a log so we can see what is going on - so now we can't see what is going on.  Please put them back.

**Pull request closed without merging**